main.py

```diff
@@ -1,53 +1,223 @@
+                    pd.DataFrame(
+                        {"loss": running_loss, "accuracy": running_accuracy, "val_loss": val_loss, "val_accuracy": val_accuracy},
+                        index=[epoch],
+                    ),
+                ]
+            )
+            print(
+                f"[{epoch + 1}] loss: {running_loss:.3f} accuracy: {running_accuracy:.3f}"
+                + f" val_loss: {val_loss:.3f} val_accuracy: {val_accuracy:.3f}"
+            )
+        else:
+            history = pd.concat([history, pd.DataFrame({"loss": running_loss, "accuracy": running_accuracy}, index=[epoch])])
+            print(f"[{epoch + 1}] loss: {running_loss:.3f} accuracy: {running_accuracy:.3f}")
+
+    print("Finished Training")
+    return model, history
+
+
+def evaluate(model, test_dataloader: DataLoader):
+    """
+    Evaluates the performance of a given model
+
+    Args:
+        model (torch.nn.Module): The trained model to be evaluated.
+        test_dataloader (DataLoader): The test dataset loader containing notes and labels in batches.
+
+    Returns:
+        Returns:
+        tuple: A tuple containing the evaluation results.
+            - val_loss (float): The average loss of the model on the test dataset.
+            - val_accuracy (float): The accuracy of the model on the test dataset, as a percentage.
+    """
+    # get test data
+    criterion = nn.CrossEntropyLoss
+    loss_sum = 0
+    correct = 0
+    total = 0
+    with torch.no_grad():
+        for data in test_dataloader:
+            notes, labels = data
+            total += labels.size(0)
+            out = model(notes)
+            loss = criterion(out, labels)
+            loss_sum += loss.item()
+            _, preds = torch.max(out, 1)
+            for pred, label in zip(preds, labels):
+                if pred == label:
+                    correct += 1
+    val_loss = loss_sum / len(test_dataloader)
+    val_accuracy = correct / total
+    return val_loss, val_accuracy
+
+
+def test_model(model: nn.Module, test_data: DataLoader, path: Optional[str] = None, classnames=None):
+    """
+    Evaluate the performance of the trained ComposerClassifier model with the provided test data loader.
+
+    Args:
+        model (PitchSeqNN): The trained ComposerClassifier model to be evaluated.
+        test_data (DataLoader): Data for the model to be tested on
+        path (str, optional): If specified, the function will load the model's state from the
+                              provided path.
+        classnames (list[str], optional): Classnames for printing results.
+    """
+    if classnames is None:
+        classnames = [0, 1]
+    # if path is specified load state from file
+    if path is not None:
+        model.load_state_dict(torch.load(path))
+    # containers for counting prediction
+    correct_pred = {classname: 0 for classname in classnames}
+    total_pred = {classname: 0 for classname in classnames}
+    correct = 0
+    total = 0
```

Comment:
```suggestion
    correct_pred = {classname: 0 for classname in classnames}
```

I think there is no need to count the "total" numbers, you can get this by doing `len(dataloader.dataset)`. One container should be enough to count the correct results as well :)